src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java

```diff
@@ -0,0 +1,374 @@
+                        }
+                        retryableRequests.add(modelId);
+                    } else {
+                        LOG.info("Unexpected failure", failure);
+                    }
+                } else {
+                    successfulRequests.put(modelId, itemResponse);
+                }
+            }
+
+            // deal with not found model
+            if (notFoundModels != null) {
+                for (EntityRequest origRequest : toProcess) {
+                    Optional<String> modelId = origRequest.getModelId();
+                    if (modelId.isPresent() && notFoundModels.contains(modelId.get())) {
+                        // submit to cold start queue
+                        entityColdStartQueue.put(origRequest);
+                    }
+                }
+            }
+
+            processCheckpointIteration(0, toProcess, successfulRequests, retryableRequests);
+        }, exception -> {
+            if (ExceptionUtil.isOverloaded(exception)) {
+                LOG.error("too many get AD model checkpoint requests or shard not avialble");
+                setCoolDownStart();
+            } else if (ExceptionUtil.isRetryAble(exception)) {
+                // retry all of them
+                super.putAll(toProcess);
+            } else {
+                LOG.error("Fail to restore models", exception);
+            }
+        });
+    }
+
+    private void processCheckpointIteration(
+        int i,
+        List<EntityFeatureRequest> toProcess,
+        Map<String, MultiGetItemResponse> successfulRequests,
+        Set<String> retryableRequests
+    ) {
+        if (i >= toProcess.size()) {
+            return;
+        }
+
+        // whether we will process next response in callbacks
+        // if false, finally will process next checkpoints
+        boolean processNextInCallBack = false;
+        try {
+            EntityFeatureRequest origRequest = toProcess.get(i);
+
+            Optional<String> modelIdOptional = origRequest.getModelId();
+            if (false == modelIdOptional.isPresent()) {
+                return;
+            }
+
+            String detectorId = origRequest.getDetectorId();
+            Entity entity = origRequest.getEntity();
+
+            String modelId = modelIdOptional.get();
+
+            MultiGetItemResponse checkpointResponse = successfulRequests.get(modelId);
+
+            if (checkpointResponse != null) {
+                // successful requests
+                Optional<Map<String, Object>> checkpointString = checkpointDao.processRawCheckpoint(checkpointResponse.getResponse());
+                if (checkpointString.isPresent()) {
+                    // checkpoint exists
+                    Optional<Entry<EntityModel, Instant>> checkpoint = checkpointDao
+                        .fromEntityModelCheckpoint(checkpointString.get(), modelId);
+
+                    if (false == checkpoint.isPresent()) {
+                        // checkpoint is too big
```

Comment:
Is this the only case where the checkpoint isnt present?

Reply:
yes it is so far